### PR TITLE
example: fungible token

### PR DESCRIPTION
**Pull request by petersalomonsen**
There are 3 highlights in this Pull Request:

- Fungible Token contract - the standard NEAR FT contract customizable with JavaScript
- QuickJS 2024-01-13 - the latest version from https://bellard.org/quickjs/
- Complete build of the C libraries in this project rather than depending on binaries from my other project ( https://github.com/petersalomonsen/quickjs-wasm-near )

# A full example of building QuickJS for integrating with Rust in WebAssembly, and then also NEAR smart contracts

Now that this project downloads [QuickJS](https://bellard.org/quickjs/), contains all the build scripts which is part of the Rust build pipeline, and compiles for WebAssembly targets including a NEAR protocol smart contract, it serves as a comprehensive and complete example of bundling QuickJS with Rust and WebAssembly. The combination of JS, Rust and NEAR protocol smart contracts, opens up for having a lightweight JavaScript customization layer on top of robust Rust implementation of standard contracts such as NFT and FT. 

# A customizable Fungible Token contract

The added JavaScript engine makes it possible to add custom transfer functions. Below is an example from the tests, that transfers a fixed amount of `2000` tokens to `alice`. When calling this function, a `transfer_id` is returned, which `alice` then can use to return excessive funds. In this example `alice` can call the `refund` function with the `transfer_id` as argument, and then the account that did the transfer to `alice` will get `1000` tokens in refund.

My intended use case of this is to be able to reserve the max needed tokens for an AI request, and then return back to the caller the unspent tokens.

```javascript
            export function transfer_2_000_to_alice() {
                    const amount = 2_000n;
                    const transfer_id = env.signer_account_id() + '_' + new Date().getTime();
                    env.set_data(transfer_id, JSON.stringify({receiver_id: env.signer_account_id(), refund_amount: (amount / 2n).toString()}));
                    env.ft_transfer('alice.test.near', amount.toString());
                    env.value_return(transfer_id);
                }

                export function refund() {
                    const { transfer_id } = JSON.parse(env.input());
                    const {refund_amount, receiver_id} = JSON.parse(env.get_data(transfer_id));
                    env.clear_data(transfer_id);
                    env.ft_transfer(receiver_id, refund_amount);
                }
```

### CODE REVIEW

**Comment by petersalomonsen** [comment]
@race-of-sloths include

**Comment by petersalomonsen** [comment]
Application for joining @race-of-sloths can be found here: https://github.com/NEAR-DevHub/race-of-sloths/issues/187

**Comment by petersalomonsen** [comment]
@race-of-sloths include

**Comment by Tguntenaar** [comment]
@race-of-sloths score 13

File: src/wasimock.rs
Comment (Tguntenaar) [inline src/wasimock.rs:107]: Just wanted to bring this to your attention, did you leave these commented on purpose?

**Review by Tguntenaar — APPROVED** [review]
Awesome magic!

File: src/wasimock.rs
Comment (petersalomonsen) [reply src/wasimock.rs:107]: Good catch ! Yes I had to leave these commented on purpose, because otherwise I got problems when building with Rust 1.81.0. The issue is brought up here: https://github.com/rust-lang/rust/issues/130210

**Comment by race-of-sloths** [comment]
@petersalomonsen Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
King of Sloths shakes your hand! Yellow leader's jersey is yours deservedly!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://race-of-sloths.com/images/bot-banner-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://race-of-sloths.com/images/bot-banner-light.svg">
    Shows inviting banner with latest news.
</picture>](https://race-of-sloths.com/latest-news)

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=petersalomonsen/quickjs-rust-near/14&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/petersalomonsen?type=bot&pr=petersalomonsen/quickjs-rust-near/14&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/petersalomonsen)


<details><summary>Current status: <i>executed</i></summary>

| Reviewer | Score |
|--------|--------|
| @Tguntenaar  | 13 |


Your contribution is much appreciated with a final score of 13!
You have received 137 (130 base + 5% lifetime bonus) Sloth points for this contribution


We would appreciate your feedback or contribution idea that you can submit [here](https://github.com/NEAR-DevHub/race-of-sloths/issues/new/choose)
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/petersalomonsen)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Pull request merged**
